packages/http/src/router/matchBaseUrl.ts

```diff
@@ -1,23 +1,55 @@
+import * as E from 'fp-ts/lib/Either';
+import * as O from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+
 import { INodeVariable, IServer } from '@stoplight/types';
 import { MatchType } from './types';
 
 const variableRegexp = /{(.*?)}/g;
 
-export function matchBaseUrl(server: IServer, baseUrl: string): MatchType {
-  const templateMatchResult = matchRequestUrlToTemplateUrl(baseUrl, server.url, server.variables);
+export function matchBaseUrl(server: IServer, baseUrl: string) {
+  return pipe(
+    convertTemplateToRegExp(server.url, server.variables),
+    E.map(regex =>
+      pipe(
+        O.fromNullable(regex.exec(baseUrl)),
+        O.fold(
+          () => MatchType.NOMATCH,
+          matches => (matches.length > 1 ? MatchType.TEMPLATED : MatchType.CONCRETE)
+        )
+      )
+    ),
+    E.fold(
+      e => {
+        throw e;
+      },
+      matchType => matchType
+    )
+  );
+}
 
-  if (!templateMatchResult) {
-    return MatchType.NOMATCH;
-  }
+export function convertTemplateToRegExp(
+  urlTemplate: string,
+  variables?: { [name: string]: INodeVariable }
+): E.Either<Error, RegExp> {
+  return pipe(
+    O.fromNullable(variables),
+    O.fold(
+      () => E.right(urlTemplate),
+      vars => replaceString(vars, urlTemplate)
+    ),
+    E.map(regexString => new RegExp(`^${regexString}$`))
+  );
 
-  return templateMatchResult.length > 1 ? MatchType.TEMPLATED : MatchType.CONCRETE;
-}
+  function replaceString(vars: { [name: string]: INodeVariable }, input: string) {
+    return E.tryCatch<Error, string>(
+      () => replaceStringUnsafe(input),
+      e => (e instanceof Error ? e : new Error('unknown error'))
```

Comment:
Why wouldn't this be an error?

Reply:
This was adopted from the example in the `fp-ts` documentation. In this case I do know that it will be an `Error`, so I could really skip the check, but even then, `e` is non-generically typed as `unknown`. How do I solve it in a simpler way without a typecast?

Reply:
Using the `toError` function from fp-ts :trollface: 

```suggestion
      E.toError
```

Reply:
This library starts to feel like lodash. _There's a function for everything_

Reply:
It's battle tested, they've gone through these use cases all the times. You were right on the approach though, which is good.